scripts/caclmgrd

```diff
@@ -154,40 +154,58 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
             self.namespace_docker_mgmt_ip[namespace] = self.get_namespace_mgmt_ip(self.iptables_cmd_ns_prefix[namespace],
                                                                                              namespace)
             self.namespace_docker_mgmt_ipv6[namespace] = self.get_namespace_mgmt_ipv6(self.iptables_cmd_ns_prefix[namespace],
                                                                                              namespace)
 
     def get_namespace_mgmt_ip(self, iptable_ns_cmd_prefix, namespace):
-        ip_address_get_command = iptable_ns_cmd_prefix + "ip -4 -o addr show " + ("eth0" if namespace else "docker0") +\
-                                 " | awk '{print $4}' | cut -d'/' -f1 | head -1"
+        ip_address_cmd0 = iptable_ns_cmd_prefix + ['ip', '-4', '-o', 'addr', 'show', ("eth0" if namespace else "docker0")]
+        ip_address_cmd1 = ['awk', '{print $4}']
+        ip_address_cmd2 = ['cut', '-d', '/', '-f1']
+        ip_address_cmd3 = ['head', '-1']
 
-        return self.run_commands([ip_address_get_command])
+        return self.run_commands_pipe(ip_address_cmd0, ip_address_cmd1, ip_address_cmd2, ip_address_cmd3)
 
     def get_namespace_mgmt_ipv6(self, iptable_ns_cmd_prefix, namespace):
-        ipv6_address_get_command = iptable_ns_cmd_prefix + "ip -6 -o addr show scope global " + ("eth0" if namespace else "docker0") +\
-                                 " | awk '{print $4}' | cut -d'/' -f1 | head -1"
-        return self.run_commands([ipv6_address_get_command])
+        ipv6_address_cmd0 = iptable_ns_cmd_prefix + ['ip', '-6', '-o', 'addr', 'show', 'scope', 'global', ("eth0" if namespace else "docker0")]
+        ipv6_address_cmd1 = ['awk', '{print $4}']
+        ipv6_address_cmd2 = ['cut', '-d', '/', '-f1']
+        ipv6_address_cmd3 = ['head', '-1']
+        return self.run_commands_pipe(ipv6_address_cmd0, ipv6_address_cmd1, ipv6_address_cmd2, ipv6_address_cmd3)
+
+    def log_output(self, cmd, exitcodes, stdout):
+        if any(exitcodes):
+            self.log_error("Error running command '{}'".format(cmd))
+        elif stdout:
+            return stdout.rstrip('\n')
 
     def run_commands(self, commands):
         """
         Given a list of shell commands, run them in order
         Args:
-            commands: List of strings, each string is a shell command
+            commands: List of List of Strings, each string is a shell command
         """
         for cmd in commands:
-            proc = subprocess.Popen(cmd, shell=True, universal_newlines=True, stdout=subprocess.PIPE)
+            proc = subprocess.Popen(cmd, universal_newlines=True, stdout=subprocess.PIPE)
 
             (stdout, stderr) = proc.communicate()
-
-            if proc.returncode != 0:
-                self.log_error("Error running command '{}'".format(cmd))
-            elif stdout:
-                return stdout.rstrip('\n')
+            self.log_output(cmd, [proc.returncode], stdout)
```

Comment:
>log_output [](http://example.com/codeflow?start=17&length=10)

This extracted function `log_output` change the original behavior: `return` will terminate for loop immediately. you new code will always return empty string in the outside function.

Reply:
You are allowed not to extract a function in this PR, I only see 2 use cases.